priam/src/main/java/com/netflix/priam/backupv2/LocalDBReaderWriter.java

```diff
@@ -0,0 +1,245 @@
+/*
+ * Copyright 2018 Netflix, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package com.netflix.priam.backupv2;
+
+import com.google.inject.Inject;
+import com.netflix.priam.config.IConfiguration;
+import com.netflix.priam.utils.GsonJsonSerializer;
+import com.netflix.priam.utils.RetryableCallable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.FileReader;
+import java.io.FileWriter;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * This class is used to create a local DB entry for SSTable components. This will be used to identify when a version
+ * of a SSTable component was last uploaded or last referenced.
+ * This db entry will be used to enforce the TTL of a version of a SSTable component as we will not be relying on
+ * backup file system level TTL.
+ * Local DB should be copied over to new instance replacing this token in case of instance replacement. If no local DB
```

Comment:
Since we're going to need to have an operation that reconstitutes the localdb anyway, I think we should take a Crash Only Software approach and treat reconstitution as the primary operation, with the localdb functioning only as a cache. In that case, we could omit the on-disk component of the localdb entirely, and just keep the relevant information in memory.

All of the information in the localdb is easily gotten from S3, with the only exception being the timeLastReferenced. For that, we could take one of two approaches:
* diff manifests against each other, and determine timeLastReferenced from the first manifest where a given SSTable does not appear (although this is probably best renamed to firstNotReferenced)
* or do the above, and also add tombstone entries in S3 when we notice an sstable has disappeared. Deletion is then super easy: scan over objects and delete any whose tombstones are older than the TTL.

Since we have to do one of the above options anyway, I think getting rid of the on-disk component makes operations (node replacement) easier and reduces complexity within Priam.

Reply:
1. Every priam startup then would have to download all the meta files and scan over 100's of thousands of SSTables (via list operation). List operation is time-consuming and I prefer instant startups. 
2. I don't like creating a strong dependency on a remote file system being available on startup. 

We can though avoid above by doing in the background with the best effort with failback being upload everything (an expensive operation). 

In "replacement" nodes we can land in a similar situation where the remote database is not available but in this situation, we have to upload everything (as it is a new node - new SSTables). 

To ensure a proper cleanup we will always require a background thread where we reconstitute DB so we can catch up if the remote file system is available at later time. Thus, I don't get your point of reducing "complexity"

Reply:
I'm confused, don't we just need to download the single latest "localdb" file? We may re-do some work that has been already been done as we could get stale data but that seems fine?

Like I guess what I'm saying is whatever state this file is storing, we should just put it in S3 and then we can get the entire state (possibly slightly stale) from S3 in a single GET call. I think I agree with Josh that this way we actually practice recovery (not having this file) all the time, and ensure that the metadata we're storing in S3 is fast to reconstruct on machine failure.

Reply:
In current design "localdb" is created per SSTable i.e. it contains only its components. We do not upload this localdb at all to S3 (or we did not intend to as such) as it will change almost every time we do "full" snapshot (every hour in the new world) until it gets compacted out. Note that localdb contains the time a file was last referenced to implement the TTL. 

If I understand correctly, you propose to have a single "localdb" file on disk and upload that file to S3 at regular intervals. For recovery/restart, one should download that file and then do "list" and download of every meta to implement TTL. The worst that would happen during this period is priam uploading the same file again (which for the restart can be avoided by verifying remote file system)

Another approach could be pushing the "localdb" to a real database. We intentionally avoided dependency on another database for disaster recovery.